In [27]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score





In [7]:
df = pd.read_csv(r"C:\Users\dell\OneDrive\Desktop\Excle_assignment\assignment_news\Neural networks\Alphabets_data.csv")
df.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [8]:
# ●	Begin by loading and exploring the "Alphabets_data.csv" dataset.
#  Summarize its key features such as the number of samples, features, and classes.

# shape 
df.shape

# null values
print(df.isnull().sum())


letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64


In [9]:
for col in df.columns:
    print("-----------------------")
    print(df[col].unique())
    

-----------------------
['T' 'I' 'D' 'N' 'G' 'S' 'B' 'A' 'J' 'M' 'X' 'O' 'R' 'F' 'C' 'H' 'W' 'L'
 'P' 'E' 'V' 'Y' 'Q' 'U' 'K' 'Z']
-----------------------
[ 2  5  4  7  1 11  3  6 12  8  9 10 13  0 15 14]
-----------------------
[ 8 12 11  1  2 15  9 13  4 10  3  5  6 14  7  0]
-----------------------
[ 3  6  5  4 13  8  2  7 11 12  9  1 10  0 15 14]
-----------------------
[ 5  7  8  6  1  4  2  9  3 11 10  0 12 13 15 14]
-----------------------
[ 1  2  6  3  4  7  9  5 10  8  0 11 14 13 12 15]
-----------------------
[ 8 10  5 13  6  7  3  9  0  4  1 12 11 14  2 15]
-----------------------
[13  5  6  9  8  7  2 11 10  1  4 14  0  3 12 15]
-----------------------
[ 0  5  2  4  6  3  7  9  1  8 10 13 14 15 12 11]
-----------------------
[ 6  4  9  2  8  3  5 11  7  1 10  0 13 14 12 15]
-----------------------
[ 6 13 10  4  5  7  8 12 11 14  0  9  3 15  1  2]
-----------------------
[10  3  4  5  6  2  1  7  9  8 11  0 12 14 13 15]
-----------------------
[ 8  9  7 10  6 11  4 12 14  3 

In [10]:
df.describe()


xbox          ybox         width       height         onpix  \
count  20000.000000  20000.000000  20000.000000  20000.00000  20000.000000   
mean       4.023550      7.035500      5.121850      5.37245      3.505850   
std        1.913212      3.304555      2.014573      2.26139      2.190458   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        3.000000      5.000000      4.000000      4.00000      2.000000   
50%        4.000000      7.000000      5.000000      6.00000      3.000000   
75%        5.000000      9.000000      6.000000      7.00000      5.000000   
max       15.000000     15.000000     15.000000     15.00000     15.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.897600      7.500450      4.628600      5.178650      8.282050   
std        2.026035      2.325354      2.699968      2.380823      2.488475   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

            x2ybar        xy2bar         xedge        xedgey         yedge  \
count  20000.00000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.45400      7.929000      3.046100      8.338850      3.691750   
std        2.63107      2.080619      2.332541      1.546722      2.567073   
min        0.00000      0.000000      0.000000      0.000000      0.000000   
25%        5.00000      7.000000      1.000000      8.000000      2.000000   
50%        6.00000      8.000000      3.000000      8.000000      3.000000   
75%        8.00000      9.000000      4.000000      9.000000      5.000000   
max       15.00000     15.000000     15.000000     15.000000     15.000000   

            yedgex  
count  20000.00000  
mean       7.80120  
std        1.61747  
min        0.00000  
25%        7.00000  
50%        8.00000  
75%        9.00000  
max       15.00000

In [11]:
# Target column (alphabet classes)
df.head()

print(df['letter'].nunique())
print(df['letter'].value_counts())

26
letter
U    813
D    805
P    803
T    796
M    792
A    789
X    787
Y    786
N    783
Q    783
F    775
G    773
E    768
B    766
V    764
L    761
R    758
I    755
O    753
W    752
S    748
J    747
K    739
C    736
H    734
Z    734
Name: count, dtype: int64


In [12]:
df[df['letter']=="U"]

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
99         U     7    11      8       9      4     3     9      6      7   
115        U     4     7      4       5      2     7     5     13      5   
125        U     5     6      5       4      2     5     8      6      8   
133        U     3     3      3       1      1     5     8      5      7   
149        U     5     7      6       5      5     8     8      8      5   
...      ...   ...   ...    ...     ...    ...   ...   ...    ...    ...   
19838      U     6    11      6       6      4     6     5      6      5   
19876      U     4     5      5       4      3     6     8      5      7   
19909      U     9    11      9       8      6     3     8      5      8   
19914      U     4     4      5       3      2     5     8      5      7   
19956      U     3     7      5       5      3     4     8      6      6   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
99        11      11      10      3       9      2       6  
115        7      14       7      3       9      0       8  
125        9       8       8      3       9      2       5  
133       10       9       8      3      10      2       6  
149        6       7       9      4       8      4       5  
...      ...     ...     ...    ...     ...    ...     ...  
19838      6       9       9      5       9      3      10  
19876      7      10      10      3       9      1       8  
19909     10       9      10      3       9      2       6  
19914     10       9       8      3       9      2       5  
19956      7       9      10      3       9      1       8  

[813 rows x 17 columns]

In [13]:
# ●	Execute necessary data preprocessing steps including data normalization, managing missing values.

df = df.dropna()
df



letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
0          T     2     8      3       5      1     8    13      0      6   
1          I     5    12      3       7      2    10     5      5      4   
2          D     4    11      6       8      6    10     6      2      6   
3          N     7    11      6       6      3     5     9      4      6   
4          G     2     1      3       1      1     8     6      6      6   
...      ...   ...   ...    ...     ...    ...   ...   ...    ...    ...   
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          6      10       8      0       8      0       8  
1         13       3       9      2       8      4      10  
2         10       3       7      3       7      3       9  
3          4       4      10      6      10      2       8  
4          6       5       9      1       7      5      10  
...      ...     ...     ...    ...     ...    ...     ...  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8  

[20000 rows x 17 columns]

In [14]:
# Normalization / Scaling

# Standardization (Mean = 0, Std = 1)

x = df.drop(columns="letter",axis=1)
y = df['letter']

scaler = StandardScaler()
X_scaled = scaler.fit_transform(x)

X_scaled


array([[-1.0576983 ,  0.29187713, -1.05327668, ..., -0.21908163,
        -1.4381527 ,  0.12291107],
       [ 0.51038497,  1.5023577 , -1.05327668, ..., -0.21908163,
         0.12008142,  1.35944092],
       [-0.01230945,  1.19973756,  0.43590966, ..., -0.8656262 ,
        -0.26947711,  0.74117599],
       ...,
       [ 1.03307939,  0.59449727,  0.43590966, ...,  2.36709667,
        -0.65903564, -2.35014863],
       [-1.0576983 , -1.22122359, -0.55688123, ...,  0.42746295,
         0.50963994,  0.12291107],
       [-0.01230945,  0.59449727,  0.43590966, ..., -0.8656262 ,
        -0.65903564,  0.12291107]])

In [15]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2)


2. Model Implementation

In [18]:
# Encode the Target Column (Alphabet Letters)

df['letter'] = df['letter'].astype('category').cat.codes

df.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      19     2     8      3       5      1     8    13      0      6      6   
1       8     5    12      3       7      2    10     5      5      4     13   
2       3     4    11      6       8      6    10     6      2      6     10   
3      13     7    11      6       6      3     5     9      4      6      4   
4       6     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [19]:
# Split Features (X) and Target (y)

X = df.drop('letter', axis=1)
y = df['letter']




In [20]:
# Normalize the Data (0–1 Range)
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Train–Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)




In [21]:
# Build the ANN Model (With Hidden Layer)
model = Sequential()

# Hidden Layer (32 neurons)
model.add(Dense(32, activation='relu', input_shape=(X_train.shape[1],)))

# Output Layer (26 classes)
model.add(Dense(26, activation='softmax'))



In [22]:
# Compile the Model
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)




In [23]:
# Train the Model
history = model.fit(
    X_train, y_train,
    epochs=20,
    batch_size=32,
    validation_split=0.2
)




Epoch 1/20
400/400 [==============================] - 10s 7ms/step - loss: 3.1289 - accuracy: 0.1438 - val_loss: 2.8965 - val_accuracy: 0.2700
Epoch 2/20
400/400 [==============================] - 2s 6ms/step - loss: 2.6030 - accuracy: 0.3720 - val_loss: 2.3497 - val_accuracy: 0.4356
Epoch 3/20
400/400 [==============================] - 3s 7ms/step - loss: 2.1502 - accuracy: 0.4848 - val_loss: 2.0037 - val_accuracy: 0.5325
Epoch 4/20
400/400 [==============================] - 2s 5ms/step - loss: 1.8722 - accuracy: 0.5453 - val_loss: 1.7901 - val_accuracy: 0.5684
Epoch 5/20
400/400 [==============================] - 2s 4ms/step - loss: 1.6909 - accuracy: 0.5859 - val_loss: 1.6414 - val_accuracy: 0.5819
Epoch 6/20
400/400 [==============================] - 2s 5ms/step - loss: 1.5651 - accuracy: 0.6105 - val_loss: 1.5395 - val_accuracy: 0.5916
Epoch 7/20
400/400 [==============================] - 2s 5ms/step - loss: 1.4715 - accuracy: 0.6269 - val_loss: 1.4626 - val_accuracy: 0.6369
Epoch

In [24]:
# Evaluate on Test Data
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Test Accuracy:", test_acc)



125/125 [==============================] - 1s 3ms/step - loss: 1.0483 - accuracy: 0.7330
Test Accuracy: 0.7329999804496765


3. Hyperparameter Tuning

In [26]:
# QUICK MANUAL EXPERIMENTS (fast checks)

def build_model_simple(neurons=32, activation='relu', lr=0.001, hidden_layers=1):
    model = Sequential()
    model.add(Dense(neurons, activation=activation, input_shape=(X_train.shape[1],)))
    for _ in range(hidden_layers-1):
        model.add(Dense(neurons, activation=activation))
    model.add(Dense(26, activation='softmax'))
    model.compile(optimizer=Adam(learning_rate=lr),
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    return model

results = []
configs = [
    {'hidden_layers':1,'neurons':32,'lr':1e-3},
    {'hidden_layers':2,'neurons':32,'lr':1e-3},
    {'hidden_layers':1,'neurons':64,'lr':1e-3},
    {'hidden_layers':1,'neurons':32,'lr':1e-2},
]

for cfg in configs:
    model = build_model_simple(**cfg)
    history = model.fit(X_train, y_train, epochs=20, batch_size=32, validation_split=0.2, verbose=0)
    loss, acc = model.evaluate(X_test, y_test, verbose=0)
    results.append({**cfg, 'test_acc':acc})
    print(cfg, "-> test_acc:", acc)

df_results = pd.DataFrame(results)
print(df_results)



{'hidden_layers': 1, 'neurons': 32, 'lr': 0.001} -> test_acc: 0.7434999942779541
{'hidden_layers': 2, 'neurons': 32, 'lr': 0.001} -> test_acc: 0.7587500214576721
{'hidden_layers': 1, 'neurons': 64, 'lr': 0.001} -> test_acc: 0.7647500038146973
{'hidden_layers': 1, 'neurons': 32, 'lr': 0.01} -> test_acc: 0.8134999871253967
   hidden_layers  neurons     lr  test_acc
0              1       32  0.001   0.74350
1              2       32  0.001   0.75875
2              1       64  0.001   0.76475
3              1       32  0.010   0.81350


4. Evaluation

In [29]:
# Evaluation Metrics (default model)

y_pred_default = np.argmax(model.predict(X_test), axis=1)

acc_d = accuracy_score(y_test, y_pred_default)
prec_d = precision_score(y_test, y_pred_default, average='weighted')
rec_d = recall_score(y_test, y_pred_default, average='weighted')
f1_d = f1_score(y_test, y_pred_default, average='weighted')

print("DEFAULT MODEL METRICS:")
print("Accuracy:", acc_d)
print("Precision:", prec_d)
print("Recall:", rec_d)
print("F1 Score:", f1_d)



125/125 [==============================] - 21s 10ms/step
DEFAULT MODEL METRICS:
Accuracy: 0.8135
Precision: 0.8239894336935681
Recall: 0.8135
F1 Score: 0.8136892794727221


In [30]:
# Evaluation Metrics (default model)
y_pred_default = np.argmax(model.predict(X_test), axis=1)

acc_d = accuracy_score(y_test, y_pred_default)
prec_d = precision_score(y_test, y_pred_default, average='weighted')
rec_d = recall_score(y_test, y_pred_default, average='weighted')
f1_d = f1_score(y_test, y_pred_default, average='weighted')

print("DEFAULT MODEL METRICS:")
print("Accuracy:", acc_d)
print("Precision:", prec_d)
print("Recall:", rec_d)
print("F1 Score:", f1_d)


125/125 [==============================] - 3s 6ms/step
DEFAULT MODEL METRICS:
Accuracy: 0.8135
Precision: 0.8239894336935681
Recall: 0.8135
F1 Score: 0.8136892794727221


In [31]:
# Tuned Model (Hyperparameter Tuning)
model_tuned = Sequential([
    Dense(128, activation='relu', input_shape=(X_train.shape[1],)),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(len(y.unique()), activation='softmax')
])

optimizer = Adam(learning_rate=0.001)

model_tuned.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

history2 = model_tuned.fit(X_train, y_train, epochs=40, batch_size=16, validation_split=0.2)




Epoch 1/40
800/800 [==============================] - 7s 4ms/step - loss: 2.1083 - accuracy: 0.3878 - val_loss: 1.5411 - val_accuracy: 0.5547
Epoch 2/40
800/800 [==============================] - 4s 5ms/step - loss: 1.3490 - accuracy: 0.6005 - val_loss: 1.3062 - val_accuracy: 0.6231
Epoch 3/40
800/800 [==============================] - 5s 6ms/step - loss: 1.1610 - accuracy: 0.6600 - val_loss: 1.1325 - val_accuracy: 0.6781
Epoch 4/40
800/800 [==============================] - 5s 6ms/step - loss: 1.0381 - accuracy: 0.6952 - val_loss: 1.0455 - val_accuracy: 0.7066
Epoch 5/40
800/800 [==============================] - 5s 7ms/step - loss: 0.9501 - accuracy: 0.7170 - val_loss: 0.9332 - val_accuracy: 0.7275
Epoch 6/40
800/800 [==============================] - 4s 5ms/step - loss: 0.8663 - accuracy: 0.7427 - val_loss: 0.8687 - val_accuracy: 0.7416
Epoch 7/40
800/800 [==============================] - 5s 6ms/step - loss: 0.8043 - accuracy: 0.7578 - val_loss: 0.8016 - val_accuracy: 0.7644
Epoch 

Compare & Discuss Between Default Model & Tuned Model (Based on Your Results)

Tuned Model Final Performance (From Your Output)

Final Training Accuracy: 93.05%

Final Validation Accuracy: 91.91%

Final Loss: 0.2108

Final Validation Loss: 0.2450

Model ne epoch 39/40 par best performance achieve.


Overall Accuracy Improvement

The tuned model shows a significant improvement in accuracy compared to the default model.

Default Model Accuracy: (Usually around 70–78% without tuning)

Tuned Model Accuracy: ~92% validation accuracy

Discussion:
Hyperparameter tuning improved learning stability and allowed the model to learn deeper feature patterns. As a result, validation accuracy increased by approximately 15–20%, indicating better generalization.


Loss Reduction

Default Model Loss: High (typically between 0.8–1.2)

Tuned Model Loss: 0.24 validation loss

Discussion:
Lower validation loss shows that the tuned model fits the data better and has reduced error while predicting alphabet classes.



Better Convergence Pattern

Your training log shows:

Loss continuously decreases from 2.10 → 0.21

Validation accuracy increases from 55% → 92%

No major overfitting till last epochs (val_acc remains close to training_acc)

Discussion:
This stable learning curve indicates that hyperparameters such as:

Learning rate

Dense layer sizes

Number of epochs

Batch size

Activation functions

